rp2040-hal/src/xosc.rs

```diff
@@ -0,0 +1,192 @@
+    pub fn new(dev: rp2040_pac::XOSC) -> Self {
+        CrystalOscillator {
+            device: dev,
+            state: Disabled
+        }
+    }
+
+    /// Initializes the XOSC : frequency range is set, startup delay is calculated and set.
+    pub fn initialize(self, frequency: Megahertz<u32>) -> Result<CrystalOscillator<Initialized>, Error> {
+
+        const ALLOWED_FREQUENCY_RANGE: RangeInclusive<Megahertz<u32>> = Megahertz(1)..=Megahertz(15);
+        const STABLE_DELAY: Milliseconds<u64> = Milliseconds(1_u64);
+        const DIVIDER: Fraction = Fraction::new(1, 256);
+
+        if !ALLOWED_FREQUENCY_RANGE.contains(&frequency) {
+            return Err(Error::FrequencyOutOfRange)
+        }
+
+        self.device.ctrl.write(|w| {
+            w.freq_range()._1_15mhz();
+            w
+        });
+
+        let freq_hz: Hertz<u64> = frequency.into();
+        let delay_sec: Seconds<u64> = STABLE_DELAY.into();
+
+        //let startup_delay = ((freq_hz / 1000) + 128) / 256;
+        let startup_delay = delay_sec.
+            checked_mul(freq_hz.integer()).and_then(|r|
+                 r.to_generic::<u64>(DIVIDER).ok()
+             ).
+            ok_or(Error::BadArgument)?;
+
+        self.device.startup.write(|w| unsafe {
+            w.delay().bits(*startup_delay.integer() as u16);
+            w
+        });
+
+        self.device.ctrl.write(|w| {
+            w.enable().enable();
+            w
+        });
+
+        Ok(self.transition(Initialized {
+            freq_hz: freq_hz
+        }))
+    }
+}
+
+/// A token that's given when the oscillator is stablilzed, and can be exchanged to proceed to the next stage.
+pub struct StableOscillatorToken {
+    _private : ()
+}
+
+impl CrystalOscillator<Initialized> {
+
+    /// One has to wait for the startup delay before using the oscillator, ie awaiting stablilzation of the XOSC
+    pub fn await_stabilization(&self) -> nb::Result<StableOscillatorToken, Infallible> {
+
+        if self.device.status.read().stable().bit_is_clear() {
+            return Err(WouldBlock)
+        }
+
+        Ok(StableOscillatorToken {
+            _private: ()
+        })
+    }
+
+    /// Returns the stablilzed oscillator
+    pub fn get_stable(self, _token: StableOscillatorToken) -> CrystalOscillator<Stable> {
+        let freq_hz = self.state.freq_hz;
+        self.transition(Stable {
+            freq_hz:freq_hz
+        })
+    }
+}
+
+impl CrystalOscillator<Stable> {
+
+    /// Operating frequency of the XOSC in hertz
+    pub fn operating_frequency(&self) -> Hertz<u64> {
+        self.state.freq_hz
+    }
+
+    /// Disables the XOSC
+    pub fn disable(self) -> CrystalOscillator<Disabling> {
+        self.device.ctrl.modify(|_r,w| {
+            w.enable().disable();
+            w
+        });
+
+        self.transition(Disabling)
+    }
+
+    /// Put the XOSC in dormant state
+    pub fn dormant(self) -> CrystalOscillator<Dormant> {
```

Comment:
Adding the `Dormant` mode is nice, although I would try to address the two warnings listed in the datasheet:
> WARNING: stop the PLLs before selecting dormant mode
WARNING: setup the irq before selecting dormant mode
-- Datasheet page 244

Maybe there could be a token for the IRQ? I'm not sure maybe we should not make this `pub` for now until we can make it safe?

Also there needs to be a way to get the mode from `Dormant` to `Disabled` to re-enable or free the peripheral.

Reply:
Good remarks. I implemented the Dormant state to be iso-functionnal with the C SDK re. XOSC but yeah you might be right about not exposing it right now (as I don't have an idea of how to deal with IRQs just yet)

Reply:
You could also make it `unsafe` noting the Warnings from the datasheet in the Safety section of the doc comment.